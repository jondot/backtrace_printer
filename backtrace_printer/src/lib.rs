#![feature(error_generic_member_access)]

use std::io::Write;

use btparse::Frame;
use colored::Colorize;
use regex::Regex;

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error(transparent)]
    BacktraceParsing(#[from] btparse::Error),

    #[error(transparent)]
    IO(#[from] std::io::Error),
}

pub type Result<T> = std::result::Result<T, Error>;

/// filter an `std::backtrace::Backtrace` based on file name and function symbol block lists
/// NOTE: this uses a debug-output-based parser `btparse` to understand an `std` Backtrace.
///
/// # Errors
///
/// This function will return an error if parsing a backtrace fails
pub fn filter(
    bt: &std::backtrace::Backtrace,
    name_blocklist: &[Regex],
    file_blocklist: &[Regex],
) -> Result<Vec<Frame>> {
    let bt_parsed = btparse::deserialize(bt)?;
    Ok(bt_parsed
        .frames
        .into_iter()
        .filter(|x| {
            if matches!(&x.file, Some(f) if !file_blocklist.is_empty() && file_blocklist.iter().any(|l| {
              l.is_match(f)
            })) {
                return false;
            }

            if !name_blocklist.is_empty() && name_blocklist.iter().any(|l| l.is_match(&x.function)) {
                return false;
            }
            true
        })
        .collect::<Vec<_>>())
}

/// Print a backtrace given an error.
///
/// # Errors
///
/// This function will return an error if backtrace extraction fails
pub fn print_err<W: Write>(
    writer: &mut W,
    err: impl std::error::Error,
    name_blocklist: &[Regex],
    file_blocklist: &[Regex],
) -> Result<()> {
    if let Some(bt) = backtrace(&err) {
        let frames = filter(bt, name_blocklist, file_blocklist)?;
        print_frames(writer, &frames)?;
    }
    Ok(())
}

/// Print frames that were extracted and parsed from an `std::backtrace::Backtrace`
pub fn print_frames<W: Write>(writer: &mut W, frames: &Vec<Frame>) -> Result<()> {
    for frame in frames {
        writeln!(
            writer,
            "{}{}",
            frame.file.as_ref().map_or("<no file>", |file| file),
            frame
                .line
                .as_ref()
                .map_or(String::new(), |ln| format!(":{ln}")),
        )?;
        writeln!(writer, "\t{}\n", frame.function.yellow(),)?;
    }
    Ok(())
}

/// Extract a stacktrace from an error. Uses unstable features to extract
/// because apparently, there's no clear approved API for this (yet)
pub fn backtrace(err: &impl std::error::Error) -> Option<&std::backtrace::Backtrace> {
    std::error::request_ref::<std::backtrace::Backtrace>(err)
}

#[cfg(test)]
mod tests {
    use std::backtrace::Backtrace;
    use std::fs::File;
    use std::io::{self, BufWriter};

    use super::*;

    #[derive(thiserror::Error, Debug)]
    pub enum Error {
        #[error("back")]
        Io(#[from] io::Error, Backtrace),
    }

    #[test]
    pub fn test_write() {
        std::env::set_var("RUST_BACKTRACE", "1");
        let mut buf = BufWriter::new(Vec::new());

        let err: Error = File::open("does not exist").unwrap_err().into();
        print_err(&mut buf, err, &[], &[]).unwrap();
        // Do writing here.

        let bytes = buf.into_inner().unwrap();
        let string = String::from_utf8_lossy(bytes.as_slice());
        let replacer = Regex::new("/rustc/.*?/").unwrap();
        let out = replacer.replace_all(&string, "");
        insta::assert_display_snapshot!(out);
    }
}
