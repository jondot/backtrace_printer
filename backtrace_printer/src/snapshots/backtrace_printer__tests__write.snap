---
source: backtrace_printer/src/lib.rs
expression: out
---
library/std/src/../../backtrace/src/backtrace/libunwind.rs:104
	[33mstd::backtrace_rs::backtrace::libunwind::trace[0m

library/std/src/../../backtrace/src/backtrace/mod.rs:66
	[33mstd::backtrace_rs::backtrace::trace_unsynchronized[0m

library/std/src/backtrace.rs:331
	[33mstd::backtrace::Backtrace::create[0m

./src/lib.rs:99
	[33m<backtrace_printer::tests::Error as core::convert::From<std::io::error::Error>>::from[0m

library/core/src/convert/mod.rs:757
	[33m<T as core::convert::Into<U>>::into[0m

./src/lib.rs:110
	[33mbacktrace_printer::tests::test_write[0m

./src/lib.rs:106
	[33mbacktrace_printer::tests::test_write::{{closure}}[0m

library/core/src/ops/function.rs:250
	[33mcore::ops::function::FnOnce::call_once[0m

library/core/src/ops/function.rs:250
	[33mcore::ops::function::FnOnce::call_once[0m

library/test/src/lib.rs:628
	[33mtest::__rust_begin_short_backtrace[0m

library/test/src/lib.rs:651
	[33mtest::run_test_in_process::{{closure}}[0m

library/core/src/panic/unwind_safe.rs:272
	[33m<core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once[0m

library/std/src/panicking.rs:552
	[33mstd::panicking::try::do_call[0m

library/std/src/panicking.rs:516
	[33mstd::panicking::try[0m

library/std/src/panic.rs:142
	[33mstd::panic::catch_unwind[0m

library/test/src/lib.rs:651
	[33mtest::run_test_in_process[0m

library/test/src/lib.rs:574
	[33mtest::run_test::{{closure}}[0m

library/test/src/lib.rs:602
	[33mtest::run_test::{{closure}}[0m

library/std/src/sys_common/backtrace.rs:154
	[33mstd::sys_common::backtrace::__rust_begin_short_backtrace[0m

library/std/src/thread/mod.rs:529
	[33mstd::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}[0m

library/core/src/panic/unwind_safe.rs:272
	[33m<core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once[0m

library/std/src/panicking.rs:552
	[33mstd::panicking::try::do_call[0m

library/std/src/panicking.rs:516
	[33mstd::panicking::try[0m

library/std/src/panic.rs:142
	[33mstd::panic::catch_unwind[0m

library/std/src/thread/mod.rs:528
	[33mstd::thread::Builder::spawn_unchecked_::{{closure}}[0m

library/core/src/ops/function.rs:250
	[33mcore::ops::function::FnOnce::call_once{{vtable.shim}}[0m

library/alloc/src/boxed.rs:2007
	[33m<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once[0m

library/alloc/src/boxed.rs:2007
	[33m<alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once[0m

library/std/src/sys/unix/thread.rs:108
	[33mstd::sys::unix::thread::Thread::new::thread_start[0m

<no file>
	[33m__pthread_joiner_wake[0m


